Initial Classes

mecha opened the pull request:
Initial classes for nonces, nonce factories, nonce field blocks and nonce validators.

mecha committed:
Updated dependencies

mecha committed:
Added `NonceAwareTrait`

mecha committed:
Added `AbstractNonce`

mecha committed:
Added `Nonce`

mecha committed:
Added `AbstractNonceValidator`

mecha committed:
Updated PHP CS Fixer for PHP v5.4

mecha committed:
Fixed code standards

mecha committed:
Added mocked WP functions file for tests

mecha committed:
`AbstractNonce` code can no longer be null

mecha committed:
Updated `NonceAwareTrait` description

mecha committed:
Added `AbstractNonceFactory`

mecha committed:
Added stub for `wp_nonce_field()`

mecha committed:
Added `AbstractNonceFieldBlock`

mecha committed:
Mocked wordpress functions are more accurate

mecha committed:
Added `NonceFieldBlock`

mecha committed:
Added `NonceValidator`

mecha committed:
Added mocked `wp_verify_nonce()` function

mecha committed:
Added `NonceFactory`

mecha committed:
Added string cast test for `NonceFieldBlock`

mecha committed:
Added `DelegatingNonceFactory`

mecha committed:
Added classname constant in `NonceFactoryTest`

mecha committed:
Updated `dhii/factory-interface` dependency

XedinUnknown commented on src/Block/AbstractNonceFieldBlock.php:
```diff
@@ -0,0 +1,107 @@
+    protected function _render()
+    {
+        return \wp_nonce_field(
+            $this->_getNonce()->getId(),
```
If the `NonceInterface` instance represents a nonce, and the purpose of this class is to output it, then why does it deal with some nonce ID, which isn't part of any nonce definition? If it isn't dealing with the actual nonce code, why does it need a `NonceInterface` instance?

mecha replied:
So, the output _must_ consist of the nonce code. You're right in asking why this deals with the instance, particularly because WordPress' functions are not granular enough to allow us to pass the ID (or action) and the code to a render function.

Instead, the `wp_nonce_field()` function will re-determine the code and use it for its output. I decided to use this function to keep rendering logic WordPress' responsibility. But in truth, I could have created my own rendering here, that outputs the hidden `input` field.

XedinUnknown replied:
Decided: Not using the nonce code is a bug, and the fix is to use our own rendering to make sure that the output has the actual nonce code.

XedinUnknown commented on src/Factory/AbstractNonceFactory.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    protected function _make($id, $args = [])
+    {
+        $code = $this->_generateNonceCode($id);
```
A nonce ID is not something that is part of the definition of a nonce. In WordPress, there is the concept of nonce ID (action context), but it is optional. Why isn't it at least optional here, if it needs to be there at all?

mecha replied:
The nonce ID is the action context that is used by WordPress.

I didn't really consider the optional nature of the action context, mostly because WordPress would generate the same code if asked to do so multiple times without an action context. This is also true for identical contexts.

But regardless, that's not a reason to limit things. Do you think it should optional?

XedinUnknown commented on src/Factory/DelegatingNonceFactory.php:
```diff
@@ -0,0 +1,139 @@
+     */
+    public function make($id, array $config = [])
+    {
+        return $this->_make($id, $config);
```
If the `_createNonce()` method actually makes the nonce instance, why does it use another method?

mecha replied:
The `_make` method is expected to create the entire nonce, that is, its data and the instance for that data. The `_createNonce` method is expected to create only the instance with given data.

XedinUnknown commented on src/Factory/AbstractNonceFactory.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    protected function _generateNonceCode($id)
+    {
+        return \wp_create_nonce($id);
```
If nonce factories are going to delegate creation of nonces to generic factories, why does there need to be a specialized nonce factory? Doesn't this de-centralize nonce creation?

mecha replied:
If by "specialized nonce factory" you are referring to `NonceFactory`, then the answer is that I simply included both in the repository for convenience's and completeness' sake.

XedinUnknown commented on test/unit/Validator/NonceValidatorTest.php:
```diff
@@ -0,0 +1,102 @@
+        $nonce = $this->createNonce($id, $code);
+        $subject = new NonceValidator();
+
+        $subject->validate($nonce);
```
This does not need to test that an actual nonce is validated. It only needs to test that if `wp_verify_nonce()` returns `true`, then the validation passes, and that it doesn't pass otherwise. For this, you need to mock the `wp_verify_nonce()`  function. This renders the stub functions obsolete.

XedinUnknown commented on test/functional/Block/AbstractNonceFieldBlockTest.php:
```diff
@@ -0,0 +1,127 @@
+ *
+ * @since [*next-version*]
+ */
+class sup extends TestCase
```
Class what?? :D

mecha replied:
Haha! Templating error :P I will rename.

XedinUnknown commented on test/functional/Factory/AbstractNonceFactoryTest.php:
```diff
@@ -0,0 +1,101 @@
+            'Created instance is not a valid nonce.'
+        );
+
+        $this->assertEquals($id, $nonce->getId(), 'Created nonce does not have the given ID.');
```
Why not testing the nonce code? Isn't that the most important part?

mecha replied:
Well, the factory is expected to create a nonce instance. I expect that nonce to be what I asked it would be, which only includes the ID. The nonce code is unknown is can be anything.

I can compare the generate code to the result of a manual call to `wp_create_nonce()`, if you think it's necessary.

mecha committed:
Fixed class name for `AbstractNonceFieldBlockTest`

mecha committed:
Updated repository URLs

mecha committed:
Updated dependencies

mecha committed:
`AbstractNonceFieldBlock#render()` no longer ...

... relies on WordPress' nonce rendering function.

mecha committed:
Added WP_Mock as a dev dependency

mecha committed:
`NonceFieldBlock` provides the referer URL in ...

... the same way that WordPress does. The test has also been updated to mock the required functions, rather than use the stub functions.

mecha committed:
ID parameter is optional for nonce factories

mecha committed:
Renamed `AbstractNonceFactory#_createNonce()` ...

... to `AbstractNonceFactory#_createNonceInstance()`

mecha committed:
Concrete factotry classes now implement ...

... `NonceFactoryInterface` instead of `Dhii\Factory\FactoryInterface`

mecha committed:
Tests for nonce factories now properly mock ...

... WordPress functions for creating nonces.

mecha committed:
`NonceValidatorTest` now properly mocks ...

... WordPress functions rather than using the stub functions.

mecha committed:
Removed the WordPress stub functions file

XedinUnknown commented on src/Factory/AbstractNonceFactory.php:
```diff
@@ -16,16 +16,16 @@
      * @return NonceInterface
      */
-    protected function _make($id, $args = [])
+    protected function _make($id = '', $args = [])
```
See [here](https://github.com/RebelCode/wp-nonce-interface/pull/1/commits/cd1f99cb6ec02084cc8825b7ca62ae6e5a456eca#r126361431)

mecha committed:
Updated dependencies

mecha committed:
Renamed key constants in `DelegatingNonceFactory`

mecha committed:
Factory "make" methods only accept a `$config` param

mecha committed:
Updated tests for factories

XedinUnknown commented on src/Block/NonceFieldBlock.php:
```diff
@@ -36,7 +41,18 @@ public function __construct(
+     */
+    protected function _getRefererUrl()
+    {
+        return \esc_attr(\wp_unslash($_SERVER['REQUEST_URI']));
```
`esc_attr()` should only be used where you are using it as an attribute. This method does not need to know about how it's being used. Also, if you wanna do proper WP VIP,  you should at least use [`filter_input()`] with `$filter` left default. Ideally, of course, you would use something like [PSR-7], but at this point it feels overly complicated and unfeasible.

[`filter_input()`]:              http://php.net/manual/en/function.filter-input.php
[PSR-7]:                          http://www.php-fig.org/psr/psr-7/

XedinUnknown replied:
Thinking about it, maybe it's better to have the referrer URL being injected into the block.

mecha replied:
@XedinUnknown My bad. I copy-pasted the entire thing out of WordPress without much thought. I'll remove the `esc_attr` call and make sure it is used during rendering.

Regarding the referer URL, I would tend towards injecting it.

mecha replied:
@XedinUnknown 

What do you think about removing the `$useReferer` flag from `AbstractNonceFieldBlock` and adding a `$referer` property, which if `null` results in the block not using the referer?

It makes things more concise and the constructor for the `NonceFieldBlock` cleaner by simply having the `$referer` default to `null`.

XedinUnknown replied:
I think that actually, this block is a specialization of a conceptual generic nonce block. So, yea, you can do that, but then add the WP layer.

mecha committed:
`NonceFieldBlock` requires a referer URL

mecha committed:
Changed `AbstractNonceFieldBlock`'s `useReferer` ...

... flag to a `$refererUrl` string which disables the referer when `null`.

mecha committed:
WordPress' `esc_attr()` function now used in rendering

mecha committed:
Added function mocking for `esc_attr()`

XedinUnknown commented on src/Factory/DelegatingNonceFactory.php:
```diff
@@ -0,0 +1,139 @@
+     *
+     * @since [*next-version*]
+     */
+    const K_DELEGATE_CONFIG_ID = 'id';
```
Already exists in ancestor, I think.

mecha replied:
It's called differently: `K_CONFIG_ID`, the key of the ID in the config passed to a nonce factory. The `K_DELEGATE_*` constants are keys used when passing config to a generic factory.

XedinUnknown commented on src/Block/AbstractNonceFieldBlock.php:
```diff
@@ -0,0 +1,154 @@
+     *
+     * @return string
+     */
+    protected function _render()
```
Ideally, this would be abstract, and the real logic would exist in the concrete implementation.

mecha replied:
Unsurprisingly, I agree with you. Surprisingly, I didn't do this from the start.

mecha committed:
`AbstractNonceFieldBlock#_render()` is now abstract ...

and rendering logic was moved to `NonceFieldBlock#_render()`.

XedinUnknown reviewed (approved):
Great, now we have all the WP Scheiße in one place.

mecha committed:
Updated dependencies
